share/availability/discovery/discovery.go

```diff
@@ -317,46 +308,51 @@ func (d *Discovery) findPeers(ctx context.Context) bool {
 	}
 }
 
-// Advertise is a utility function that persistently advertises a service through an Advertiser.
-// TODO: Start advertising only after the reachability is confirmed by AutoNAT
-func (d *Discovery) Advertise(ctx context.Context) {
-	if d.params.AdvertiseInterval == -1 {
-		return
+// handleDiscoveredPeer adds peer to the internal if can connect or is connected.
+// Report whether it succeeded.
+func (d *Discovery) handleDiscoveredPeer(ctx context.Context, peer peer.AddrInfo) bool {
+	logger := log.With("peer", peer.ID)
+	switch {
+	case peer.ID == d.host.ID():
+		logger.Debug("skip handle: self discovery")
+		return false
+	case len(peer.Addrs) == 0:
+		logger.Debug("skip handle: empty address list")
+		return false
+	case d.set.Size() >= d.set.Limit():
+		logger.Debug("skip handle: enough peers found")
+		return false
+	case d.connector.HasBackoff(peer.ID):
+		logger.Debug("skip handle: backoff")
+		return false
 	}
 
-	timer := time.NewTimer(d.params.AdvertiseInterval)
-	defer timer.Stop()
-	for {
-		ttl, err := d.disc.Advertise(ctx, topic)
+	switch d.host.Network().Connectedness(peer.ID) {
+	case network.Connected:
+		d.connector.Backoff(peer.ID) // we still have to backoff the connected peer
+	case network.NotConnected:
+		err := d.connector.Connect(ctx, peer)
 		if err != nil {
-			log.Debugf("Error advertising %s: %s", topic, err.Error())
-			if ctx.Err() != nil {
-				return
-			}
-
-			select {
-			case <-timer.C:
-				timer.Reset(d.params.AdvertiseInterval)
-				continue
-			case <-ctx.Done():
-				return
-			}
+			logger.Debugw("unable to connect", "err", err)
+			return false
 		}
+	default:
+		panic("unknown connectedness")
```

Comment:
aren't there two more kinds of connectedness ? feels weird to panic here.

Reply:
We had an argument with Vlad

Reply:
There are no more types of connectedness here, but if there are introduced and we missed that, it's a correctness bug and our testing pipeline should catch the panic so we can resolve